### ISSUE TITLE
Concept for "untracable landing page"

### ISSUE THREAD

**Issue by ninoppp** [freedomofpress/securedrop#6643]
Hey, I hope this is the right place for this issue, feel free to move it somewhere else if it isn't.

While planning the implementation of our landing page, we came up with a concept for an "invisible landing page": instead of creating a whole subpage for the whistleblowing/securedrop instructions, we emulate one with a pure css popup that looks like a subpage. Very lazy example: https://codepen.io/ninoprs/pen/KKRYode

This would have some considerable benefits: 
- As long as the webpage analytics only log page vistis and not clicks, or at least not this particular click, there would be no way to trace any opening from the server side
- Nearly impossible to trace in the client's browser, not visible in browser history, not visible to browser plugins...
- Could still be made available through link via css id if needed
- As long a design is kept minimal, increase in load time should be neglectable.

Does this make sense to you, was something like this ever discussed? If you're interested I'd be happy to write a real concept, with all of the security benefits, and make some additions to the documentation.

**Comment by eaon** [comment]
I would love if there was a better way along those lines, however, I have some rather large concerns …

> - As long as the webpage analytics only log page vistis and not clicks, or at least not this particular click, there would be no way to trace any opening from the server side

Unfortunately, the existence of JavaScript on the same page would be worrying still. For one-off verification, it might just be a little more effort than just checking the webpage's source code, but continuously relying on newsroom's web departments having enough awareness of the landing page's needs strikes me as optimistic.

> - Nearly impossible to trace in the client's browser, not visible in browser history, not visible to browser plugins...

I don't think this is true, at least Firefox does enter the popup in the example you posted into the browser history :frowning_face: It would be possible to remove it deliberately with JavaScript, but … see above - it strikes me as brittle. I also very much doubt it would be invisible to browser plugins full stop, as they can typically have access to the entire DOM and state of anything that's on screen.

Wondering what @l3th3's view is on this.

**Comment by ninoppp** [comment]
Thanks a lot for the feedback. 
> Unfortunately, the existence of JavaScript on the same page would be worrying still. For one-off verification, it might just be a little more effort than just checking the webpage's source code, but continuously relying on newsroom's web departments having enough awareness of the landing page's needs strikes me as optimistic.

Well, I deffo see the point. Coming from a small, up and coming magazine where the website is being designed with the landing page in mind (as well as privacy in general, using selfhosted, limited analytics), I personally would be positive that this would be handlable. However, yeah, bigger operations will deffo be more complex.

> I don't think this is true, at least Firefox does enter the popup in the example you posted into the browser history ☹️ It would be possible to remove it deliberately with JavaScript, but … see above - it strikes me as brittle. 

Okay, I clearly don't know enough about browser logging (which is why i posted the issue xD). Could you tell me in which exact file this is visible, and maybe what practice is used for mitigatino through javascript? Also, I opened the pen with firefox for the first time and it loks really broken, is that just me?

> I also very much doubt it would be invisible to browser plugins full stop, as they can typically have access to the entire DOM and state of anything that's on screen.

Fair point, let me rephrase: It is less likely to be tracked by a browser plugin, compared to a conventional page.

Well, that mainly leaves the "too complicated" problem. A failsafe we're planing is to implement is to regularly scan for changes in the "landing page", preferably before they're life. Gonna have to figure out what the best solution is for that, maybe it's something that could be easily replicated. Actually scaning your own site for watering hole attacks might not be such a bad idea. Or is that something that's already standard practice in big media companies?

**Comment by eaon** [comment]
> … where the website is being designed with the landing page in mind (as well as privacy in general, using selfhosted, limited analytics) …

Right, unfortunately, our formal publicly documented requirements for directory inclusion have to be a bit more defensive and take into account how this type of thing typically goes wrong.

> Could you tell me in which exact file this is visible, and maybe what practice is used for mitigatino through javascript?

IIRC the file would be `places.sqlite`, but in this case I just checked the browser's own history UI during the session. As to how to mess with that history, the [`location.replace()`](https://developer.mozilla.org/en-US/docs/Web/API/Location/replace) is what you'd need.

> Also, I opened the pen with firefox for the first time and it loks really broken, is that just me?

It does look a bit broken but I don't think it affects the functionality.

> Fair point, let me rephrase: It is less likely to be tracked by a browser plugin, compared to a conventional page.

I'm not sure about that either because of the URL/history traces. I believe there may be a way to get what you're after by using `<summary>` and `<details>`, or the kind of checkbox tricks we use in the source interface with the codename/passphrase widget and locale menu - I think that would definitely need a targeted tracker and wouldn't allow for after-the-fact forensic analysis of a source's machine. But just to be clear, I would treat it as research rather than a path to a landing page to be included in the SecureDrop directory.

I wish there was a way to exclude sections of markup from being accessed / mangled by JavaScript - an inverse of `<noscript>` sort of - but I believe that is such a niche use-case that I doubt it would get traction in the standardisation bodies …

As for (self-)scanning practices, I have no knowledge of this so wouldn't be able to speak to that.

**Comment by ninoppp** [comment]
> In this case I just checked the browser's own history UI during the session. 

Wait what? You mind sending me a screenshot of the history? I only see the codepen site opened, nothing about a popup. Can't find anything in the file either, but not an expert tho.

> I believe there may be a way to get what you're after by using <summary> and <details>

That's an intersting idea. Will take a look.

> But just to be clear, I would treat it as research rather than a path to a landing page to be included in the SecureDrop directory.

I see that that may be extra work for the SecureDrop team, but does that mean you wouldn't accept an implementation of this into the directory, that, while maybe not perfect, would at least in some ways be more secure than the traditional landing page, if it's cleanly executed?

**Comment by eaon** [comment]
> Wait what? You mind sending me a screenshot of the history?

[Screencast from 2022-11-07 10:15:16.webm](https://user-images.githubusercontent.com/201369/200346504-c06fc1e7-0f4f-48af-8a72-9ad8fbb591ad.webm)

FYI, I downloaded your example and hosted it locally (I also changed the background to grey rather than an image), but functionality is the same.

> … but does that mean you wouldn't accept an implementation of this into the directory, that, while maybe not perfect, would at least in some ways be more secure than the traditional landing page, if it's cleanly executed?

That's not a decision that is down to me, but to be transparent about my own views here, I strongly prefer compliance with publicly accessible specifications over case-by-case exceptions. And at least as it stands now, I would be against making changes to the landing page requirements to allow for unrelated JavaScript on the same page. We'd be entering a _continuously trust the orgs web developer team_ realm and I don't think that's a good place to be in.

**Comment by ninoppp** [comment]
Thanks for the video.

>  I strongly prefer compliance with publicly accessible specifications over case-by-case exceptions.

Totally see that point. Guess we'll make a conventional landing page and then do some research into an untracable one and see where it leads us. 

> We'd be entering a continuously trust the orgs web developer team realm and I don't think that's a good place to be in.

Hmmm, well, you're already trusting the beckend people to keep the headers etc intact., including things that aren't visible from the outside, such as the handling of the logs (which I can think of quite a few scenarios of for a fuckup). So I would argue, that you are shifting the trust from the backend to the frontent (assuming the concept more or less works like I initially described). And that, since there are no hidden parts, it is theoretically possible for any party to verify, that everything is at it should be, eliminating the need for trust altogether. Of course the right set up would need to be in place for that.
